src/web/extractors.rs

```diff
@@ -33,26 +47,214 @@ pub struct UidParam {
+                        } => valid.push(BatchBsoBody {
+                            id,
+                            sortindex,
+                            payload,
+                            ttl,
+                        }),
+                        bso => invalid.push(bso),
+                    }
+                } else {
+                    invalid.push(bso);
+                }
+            }
+            future::ok(BsoBodies { valid, invalid })
+        });
+
+        Box::new(fut)
+    }
+}
+
+#[derive(Default, Deserialize, Serialize, Validate)]
 pub struct BsoBody {
+    #[validate(custom = "validate_body_bso_id")]
+    pub id: Option<String>,
+    #[validate(custom = "validate_body_bso_sortindex")]
     pub sortindex: Option<i32>,
+    #[validate(custom = "validate_body_bso_payload")]
     pub payload: Option<String>,
+    // No validation here as the limits are lenient
     pub ttl: Option<u32>,
 }
 
+impl FromRequest<ServerState> for BsoBody {
+    type Config = ();
+    type Result = Box<Future<Item = BsoBody, Error = Error>>;
+
+    fn from_request(req: &HttpRequest<ServerState>, _: &Self::Config) -> Self::Result {
+        // Only try and parse the body if its a valid content-type
+        let headers = req.headers();
+        let default = HeaderValue::from_static("");
+        match headers.get(CONTENT_TYPE).unwrap_or(&default).as_bytes() {
+            b"application/json" | b"text/plain" | b"" => (),
+            _ => {
+                // TODO: This is supposed to return a 415 status for unknown content-type
+                return Box::new(future::err(
+                    ValidationErrorKind::FromDetails(
+                        "Invalid content-type".to_owned(),
+                        RequestErrorLocation::Header,
+                        Some("content-type".to_owned()),
+                    ).into(),
+                ));
+            }
+        }
+        let mut config = JsonConfig::default();
+        config.limit(BSO_MAX_PAYLOAD_SIZE);
+        let fut = <Json<BsoBody>>::from_request(req, &config)
+            .map_err(|e| {
+                let err: ApiError = ValidationErrorKind::FromDetails(
+                    e.to_string(),
+                    RequestErrorLocation::Body,
+                    Some("bso".to_owned()),
+                ).into();
+                err.into()
+            }).and_then(|mut bso: Json<BsoBody>| {
+                match bso.validate() {
+                    Ok(()) => (),
+                    Err(e) => {
+                        let err: ApiError = ValidationErrorKind::FromValidationErrors(
+                            e,
+                            RequestErrorLocation::Body,
+                        ).into();
+                        return future::err(err.into());
+                    }
+                }
+
+                // Validate the TTL, comment copied from Python, bug seems resolved
+                // so this workaround should no longer be necessary, verify during PR review.
```

Comment:
Question for @rfk...?

Reply:
I talked with @rfk, he thinks it should be fine to validate it properly now, so I'll update and remove this exception.